### ISSUE TITLE
Time values for volumes.Volume are incorrectly handled

### ISSUE THREAD

**Issue by markdryan** [gophercloud/gophercloud#248]
The volume.Volumes struct has two time fields, CreatedAt and UpdatedAt.  Both of these fields are defined to be of type time.Time.  However, a separate type, gophercloud.JSONRFC3339MilliNoZ, is used to unmarshall them.  This creates two problems:

1. It's not possible to unmarshall a json representation of a volumes.Volume structure that contains a CreatedAt or UpdatedAt field with a timezone, e.g., "created_at": "2015-09-17T03:35:03.000000Z".  According to the [spec](https://developer.openstack.org/api-ref/block-storage/v3/index.html?expanded=list-accessible-volumes-detail,list-accessbile-volumes-with-details-detail#list-accessbile-volumes-with-details) this should be possible.
2. Timestamps without time zone information are processed incorrectly.  They are parsed using time.Parse which treats them as being UTC times which is not right.  Again according to the [spec](https://developer.openstack.org/api-ref/block-storage/v3/index.html?expanded=list-accessible-volumes-detail,list-accessbile-volumes-with-details-detail#list-accessbile-volumes-with-details) the timestamps are ISO8601 timestamps and ISO8601 timestamps without time zone information are in local and not UTC time.

The result of all this is, is that volume timestamps fields will almost always be incorrect, unless both client and server are in the UK and it's winter.

**Comment by markdryan** [comment]
I've created a PR that illustrates the issue and proposes a fix:

https://github.com/gophercloud/gophercloud/pull/249

The fix works by a) changing the type used to unmarshall volumes.Volume timestamps to gophercloud.JSONRFC3339Milli and b) fixing the UnmarshallJSON method for this type to handle timestamps that have timezone information and those that do not.  I chose this type as it was currently unused.  However, a new type could easily be introduced.

**Comment by jtopjian** [comment]
@markdryan I'm a little confused about this issue. Have you run into a deployment where the time parsing here is causing an issue?

From browsing the Cinder code, Cinder, the database schema, and Oslo are storing and retrieving everything as UTC. Any volumes I create are stored and retrieved as UTC. If I set my server's time zone to my local zone (MDT), all dates are still stored as UTC.

Now, I totally get that not having the `Z` at the end is incorrect ISO8601 form. Googling around, I see this is not uncommon, especially with MySQL. But I get that others might have different environments configured, which is why I'm curious if you've seen this issue first hand.

Also, I did a quick modification of [this](https://github.com/gophercloud/gophercloud/blob/master/acceptance/openstack/blockstorage/v2/volumes_test.go#L34) test to print out the volume creation time:

```go
t.Logf("Volume time: %s", newVolume.CreatedAt)
```

Without your patch in #249, I get this:

```
$ date
Sun May 21 22:53:20 MDT 2017
$ gophercloudtest TestVolumesCreateDestroy blockstorage/v2
=== RUN   TestVolumesCreateDestroy
--- PASS: TestVolumesCreateDestroy (5.17s)
        blockstorage.go:23: Attempting to create volume: ACPTTESTg02EeG7J1fQxBswN
        volumes_test.go:51: Volume time: 2017-05-22 04:53:25 +0000 UTC
```

And with your patch, I get this:

```
$ date
Sun May 21 22:54:17 MDT 2017
$ gophercloudtest TestVolumesCreateDestroy blockstorage/v2
=== RUN   TestVolumesCreateDestroy
--- PASS: TestVolumesCreateDestroy (4.64s)
        blockstorage.go:23: Attempting to create volume: ACPTTESTU67j7za2ai6S6R9D
        volumes_test.go:51: Volume time: 2017-05-22 04:54:22 -0600 MDT
```

Which isn't correct.

So I wonder if something else is a cause of the issue you're seeing?

**Comment by jtopjian** [comment]
Also, in #249, you link to [this](https://github.com/openstack/cinder/blob/master/cinder/objects/base.py#L245) piece of code. I'm assuming you meant the date fields below it (linking to master means the original link might have changed due to new code). Looking up `fields.DateTimeField` in Oslo, I see [this](https://github.com/openstack/oslo.versionedobjects/blob/a94c060e490521caa50dc2deabc2b3a9e72e6990/oslo_versionedobjects/fields.py#L467-L473) which might mean that Oslo is already accounting for the `Z`/no-`Z`-but-should-really-be-`Z` issue?

**Comment by markdryan** [comment]
@jtopjian The issue I encountered was on a project called [ciao](https://github.com/01org/ciao).  Ciao has it's own image and volume services that implement a subset of the cinder and glance apis.  We have our own command line tool, ciao-cli that uses gophercloud to talk to these services.  The problem is that the serialized form of volumes and images produced by these services contain timezone information, i.e., 'Z's in their time stamps.  When we tried to update our vendored version of gopherhcloud from rackspace/gophercloud to gophercloud/gophercloud we discovered that our command line tool could no longer parse that dates returned to it from our image and volume services.

When I investigated the problem I could see that the time parsing code in gophercloud was not handling the case where the server returned a timestamp with a trailing 'Z', so I created a patch to fix it.  I also noted that according to the specs, ISO8601 timestamps without time zone information are in local and not UTC time, so I fixed this too.  However, from your reply this second fix seems to be causing problems and is perhaps erroneous.  The code in Oslo you link to suggests that timestamps without timezone information are in fact assumed to be in UTC in Openstack.

So how about I resubmit my patch so that it just adds support for parsing image and volume timestamps that contain a trailing 'Z'?  The new patch would not change the interpretation of timestamps that do not contain a trailing Z, i.e., they would continue to be parsed as UTC times, and so shouldn't break anything.

**Comment by jtopjian** [comment]
I think that's definitely something try, and I'm happy to test it out, but I'm a little hesitant to confirm it as a solution off the bat. I think the best way to handle this is to see if it's already been solved in another SDK (perhaps even python-openstackclient or python-cinderclient, but since those are Python-based, they're probably taking advantage of the Oslo workarounds). Having some type of precedence to lean on will add a lot of confidence to the solution.

But speaking of which, are you able to interact with Ciao using the python clients? If so, how do those clients handle the time zone?

**Comment by jtopjian** [comment]
@markdryan I had a thought: If Ciao is implementing a custom version of the Glance and Cinder _services_, shouldn't it adhere to the time format of the reference (ie Glance and Cinder proper) services, warts and all?

**Comment by markdryan** [comment]
@jtopjian Sorry for my late reply.

I tried using the openstack client to query the ciao volume service as you suggested.  It works fine.

```
openstack volume show e35e6646-8481-44de-99f6-88d55773b04b                                                                                                                                                        
+------------------------------+------------------------------------------------------------+                                                                                                                                                                                   
| Field                        | Value                                                      |                                                                                                                                                                                   
+------------------------------+------------------------------------------------------------+                                                                                                                                                                                   
| bootable                     | false                                                      |                                                                                                                                                                                   
| created_at                   | 2017-05-30T11:44:44.037952676Z                             |                                                                                                                                                                                   
| encrypted                    | False                                                      |                                                                                                                                                                                   
| id                           | e35e6646-8481-44de-99f6-88d55773b04b                       |                                                                                                                                                                                   
| multiattach                  | False                                                      |                                                                                                                                                                                   
| name                         | Storage for instance: 6ef828e4-cf4f-4287-8027-7d9d2465aa77 |                                                                                                                                                                                   
| os-vol-host-attr:host        |                                                            |                                                                                                                                                                                   
| os-vol-tenant-attr:tenant_id | 80df3079e4aa4296a4c51c8eff8b8906                           |                                                                                                                                                                                   
| properties                   | contents='not junk'                                        |                                                                                                                                                                                   
| replication_status           |                                                            |                                                                                                                                                                                   
| size                         | 1                                                          |                                                                                                                                                                                   
| status                       | in-use                                                     |                                                                                                                                                                                   
| type                         | None                                                       |                                                                                                                                                                                   
| user_id                      |                                                            |                                                                                                                                                                                   
+------------------------------+------------------------------------------------------------+                          
```

When I try with our gophercloud client, I get the following error
```
./ciao-cli volume show --volume e35e6646-8481-44de-99f6-88d55773b04b                                                                                                                                              
F0530 14:01:35.882421   17769 main.go:469] ciao-cli FATAL: parsing time "2017-05-30T11:44:44.037952676Z": extra text: Z                
```

Note the error message.  It comes from https://github.com/gophercloud/gophercloud/blob/master/results.go#L236.

It would of course be possible to fix this in ciao by excluding timezone information from the date stamps produced by our volume service.  However, from our point of view it makes more sense to fix it in gophercloud as

1. There is nothing wrong with the dates produced by the ciao volume service.  They conform to the specification and it is arguably more correct to include timezone information than to not include it.
2. The openstack client tool (v 2.3.0) has no problems parsing volume dates containing timezone information.
3. Other parts of gophercloud, e.g., openstack/imageservice, have no problems processing ISO 8601 dates containing timezone information.   Gophercloud is currently a little inconsistent in the way that it handles dates.   For some services it accepts timezones for others it rejects them.
4. The old rackspace version of gophercloud (github.com/rackspace/gophercloud/) did accept volume dates with timezone information.  Something changed when the repo was moved.

It should be possible to fix the issue in a way that will not break interoperability with volume services that do not include timezone information and I'm happy to resubmit my patch to do this.  However, if you'd prefer to leave things the way they are, I'll change ciao's volume service to exclude timezone information instead.

**Comment by jrperritt** [comment]
> if you'd prefer to leave things the way they are, I'll change ciao's volume service to exclude timezone information instead.

Yes, I'd prefer this. I understand that it's probably a relatively small change and could be made in a backwards-compatible way, but that could be said for many other cases in which users have custom environments. We'd prefer to try to stick to vanilla OpenStack while allowing users to write customizations in their code.

Speaking of customizations, if it would be easier, you should be able to create a custom type, embed the `volumes.Volume`, write an `UnmarshalJSON` method for your new type, and use `volumes.commonResult.ExtractInto` to extract the data.

**Comment by markdryan** [comment]
@jrperritt Sure, no problems.  I'll fix this in ciao.  I'll also check out your suggestion to use a custom marshaler.  Thanks for the feedback and help.

I guess we can close this bug.

**Comment by jtopjian** [comment]
I've been meaning to loop back around to this one. First, I think it's really cool that you guys have implemented a custom Cinder service that is compatible with the python client. 

I do agree with @jrperritt that it's best for Ciao to modify how it's handling timestamps. From what we can see, Gophercloud is interacting with the Block Storage API correctly. But I think the fact that the python client can interact with Ciao correctly should hold some weight. I don't think any one of us here can easily explain why. For me, I'd probably have to spend a day or two digging into all sorts of Oslo libraries, poking in both server and client side code, etc, and unfortunately I don't have the time to do that. And if it turns out that the Python clients are accounting for date formats that vary from the API standard, it could be debatable whether that should be done in other implementations.

There might be a new OpenStack mailing list created specifically to discuss API issues:

https://review.openstack.org/#/c/468046/

If/when that list gets created, this might be a good topic to discuss there.

**Comment by markdryan** [comment]
From what I can tell from reading the API documents for [glance](https://developer.openstack.org/api-ref/image/v2/index.html?expanded=show-image-details-detail#show-image-details) and [cinder](https://developer.openstack.org/api-ref/block-storage/v3/index.html?expanded=list-accessible-volumes-detail,show-a-volume-s-details-detail#show-a-volume-s-details) , timestamps with and without timezone information should be accepted.

On reflection, it seems to me that gophercloud is not handling these dates correctly and that this bug should be re-opened.  For glance it accepts timestamps that contain only timezone information.  For cinder it only accepts timestamps that do not contain a timezone.  Not only is this inconsistent, given that both dates are ISO8601 dates, but it makes gophercloud very fragile.  Glance and cinder can change the format of the dates and still be compliant with the API spec.  Any such changes would break gophercloud clients.  It would not, however, break the openstack client, as we have seen.

To illustrate the problem, gophercloud cannot actually parse the example dates in the cinder API documentation.  If we place the example date 2015-08-27T09:49:58-05:00 lifted from the cinder [documentation]((https://developer.openstack.org/api-ref/block-storage/v3/index.html?expanded=list-accessible-volumes-detail,show-a-volume-s-details-detail#show-a-volume-s-details), and use it to replace a creation date used in the block storage [tests](https://github.com/gophercloud/gophercloud/blob/master/openstack/blockstorage/v2/volumes/testing/fixtures.go#L25), the tests fail.

```
convenience.go:35: Failure in requests_test.go, line 127: unexpected error "parsing time \"20\
                fmt.Fprintf(w, `                                                                     |15-08-27T09:49:58-05:00\": extra text: -05:00"   
```

This doesn't seem right.

To summarise, I think this bug needs to be re-opened, and a new patch is needed that unifies the date handling for the image and block service.  The new code would cope with timestamps that contain timezone information and those that do not.

**Comment by jtopjian** [comment]
> From what I can tell from reading the API documents for glance and cinder , timestamps with and without timezone information should be accepted.

We've seen many times in the past that the API docs aren't always correct, so we can't go by them alone.

> Not only is this inconsistent,

Unfortunately there are numerous inconsistencies between OpenStack services, so this isn't anything new or why it's not raising a flag.

>  Any such changes would break gophercloud clients

To be frank (but not critical or offensive), the only report of something breaking is from Ciao, which is a custom implementation of Cinder and Glance. All other tools that use Gophercloud have worked fine on a wide array of OpenStack deployments.

I do think there's something with regard to timezones that needs to be investigated. As I said before, the fact that the python client is handling Ciao should hold some weight. But API docs alone aren't enough to make a change. Similarly, we can't give favor to a custom implementation or else that sets a precedence for adding support for every other custom service. If someone is able to confidently explain how time is handled both in the standard service and client implementations, it would go a long way to helping solve this problem.

**Comment by markdryan** [comment]
Fair enough.  It might still make sense to re-open the bug open to represent this investigation work into timezones.  It already contains lots of useful info.

**Comment by jtopjian** [comment]
I'm OK with re-opening this until we can reach something conclusive.

**Comment by jrperritt** [comment]
As far as I can tell, no evidence has been presented to demonstrate that Gophercloud is handling this incorrectly. Because of that, the issue title is misleading. We can re-open this if someone can provide evidence in the affirmative. All the useful information in the issue will remain in tact.

**Comment by akskap** [comment]
Hi All,

I am trying to use gophercloud SDK for fetching objectstore information from two different Openstack Deployments (Suse Cloud 6 and Suse Cloud 7) and hitting the same problem as mentioned in this issue.

I was using the below code which worked fine in SuseCloud6 deployment for fetching all objects in a container:

```
		pager := objects.List(objectStorageClient, containerName, listOpts)
		page, err := pager.AllPages()
		if err != nil {
			fmt.Println("Error while calling AllPages method", err)
		}

		allObjects, err := objects.ExtractInfo(page)

```

but started failing in SuseCloud7 with the time parsing error due to different structure. As suggested in by @jrperritt  [comment here](https://github.com/gophercloud/gophercloud/issues/248#issuecomment-304928527), I tried adding/replacing the code above as follows:

Added in gophercloud/openstack/objectstorage/v1/objects/results.go
```
type OpenstackSOC7Object struct {
        Object
}

func (r *OpenstackSOC7Object) UnmarshalJSON(b []byte) error {
        fmt.Println("SOC7 unmarshal called")
        type tmp OpenstackSOC7Object
        fmt.Println("-> reached here")
        var s *struct {
                tmp
                LastModified gophercloud.JSONRFC3339Milli `json:"last_modified"`
        }

        err := json.Unmarshal(b, &s)
        if err != nil {
                return err
        }

        *r = OpenstackSOC7Object(s.tmp)

        r.LastModified = time.Time(s.LastModified)

        return nil

}

func ExtractSOC7Info(r pagination.Page) ([]OpenstackSOC7Object, error) {
        var s []OpenstackSOC7Object
        err := (r.(ObjectPage)).ExtractInto(&s)
        return s, err
}
```
and modified the previous object fetch code as below:

```
		pager := objects.List(objectStorageClient, containerName, listOpts)
		page, err := pager.AllPages()
		if err != nil {
			fmt.Println("Error while calling AllPages method", err)
		}

		allObjects, err := objects.ExtractSOC7Info(page)

```

but I am still hitting the same error. It looks like Unmarshal() is still being called for the SDK's `type Object` rather than `type OpenstackSOC7Object` I am not really clear on where/how to hook any custom code/logic in the existing SDK making sure that everything works. I also need to use the existing SDK code as it works fine for older SuseCloud6 Openstack deployments

Any help around this will be appreciated.

**Comment by akskap** [comment]
@jtopjian If possible, can you please have a quick glance on what I tried and if there is something visibly wrong with the implementation ?
